elastic/synthetics-recorder
# fix: Hoist page objects in generated code

## Summary

Resolves #195.

Resolves #185.

Depends on #205.

## Implementation details

Copies generator code from synthetics to recorder. This is done so we can modify it to fix a defect that can cause a script to reference out-of-scope vars.

## How to validate this change

You can see steps on how to re-create this problem in the linked issue. Ensure that you cannot create code that runs into this problem, and that the code we generate continues to work as designed.

Some of the code in this PR is copied directly from Synthetics, other bits are new to this change. Particular code areas to check out:

- the [function](https://github.com/elastic/synthetics-recorder/pull/211/files#diff-7c0276f32ead59b6dfb141ee9ee1426f428d8528531be9ec9f9383cae0bec593R353) that identifies the vars to hoist
- [assignment](https://github.com/elastic/synthetics-recorder/pull/211/files#diff-7c0276f32ead59b6dfb141ee9ee1426f428d8528531be9ec9f9383cae0bec593R311) of the generator's vars to hoist
- [updated conditions](https://github.com/elastic/synthetics-recorder/pull/211/files#diff-7c0276f32ead59b6dfb141ee9ee1426f428d8528531be9ec9f9383cae0bec593R204) in `generateAction` to account for whether a var is hoisted during its definition or not
- [updated logic](https://github.com/elastic/synthetics-recorder/pull/211/files#diff-7c0276f32ead59b6dfb141ee9ee1426f428d8528531be9ec9f9383cae0bec593R235) for finishing a promise all emit procedure
